packages/fxa-auth-server/lib/subhub.js

```diff
@@ -0,0 +1,315 @@
+          log.error('subhub.cancelSubscription.1', { uid, sub_id, err });
+          // TODO: update with subhub cancelSubscription error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
+          }
+          // TODO: update with subhub cancelSubscription error response for invalid plan ID
+          if (err.message === 'invalid subscription id') {
+            throw error.unknownSubscription(sub_id);
+          }
+        }
+        throw err;
+      }
+    },
+
+    async getCustomer(uid) {
+      try {
+        return await api.getCustomer(uid);
+      } catch (err) {
+        if (err.statusCode === 404) {
+          log.error('subhub.getCustomer.1', { uid, err });
+          // TODO: update with subhub createSubscription error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
+          }
+        }
+        throw err;
+      }
+    },
+
+    async updateCustomer(uid, pmt_token) {
+      try {
+        return await api.updateCustomer(uid, { pmt_token });
+      } catch (err) {
+        if (err.statusCode === 400 || err.statusCode === 404) {
+          log.error('subhub.updateCustomer.1', { uid, pmt_token, err });
+          // TODO: update with subhub createSubscription error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
+          }
+          // TODO: update with subhub updateCustomer error response for invalid payment token
+          if (err.message === 'invalid payment token') {
+            throw error.rejectedSubscriptionPaymentToken(pmt_token);
+          }
+        }
+        throw err;
+      }
+    },
+  };
+};
+
+function buildStubAPI(log, config) {
+  const {
+    subhub: {
+      stubs: {
+        plans = []
+      } = {}
+    } = {}
+  } = config;
+
+  const getPlanById = plan_id => plans
+    .filter(plan => plan.plan_id === plan_id)[0];
+
+  const storage = { subscriptions: {} };
+  const subscriptionsKey = (uid, sub_id) => `${uid}|${sub_id}`;
+
+  return {
+    isStubAPI: true,
+
+    async listPlans() {
+      return plans;
+    },
+
+    async listSubscriptions(uid) {
+      return Object
+        .values(storage.subscriptions)
+        .filter(subscription => subscription.uid === uid);
+    },
+
+    async createSubscription(uid, pmt_token, plan_id, email) {
+      const plan = getPlanById(plan_id);
+      if (! plan) {
+        throw error.unknownSubscriptionPlan(plan_id);
+      }
+      const product_id = plan.product_id;
+      const sub_id = `sub${Math.random()}`;
+      const key = subscriptionsKey(uid, sub_id);
+      storage.subscriptions[key] = {
+        uid,
+        plan_id,
+        product_id,
+        email
+      };
+      return { sub_id };
+    },
+
+    async cancelSubscription(uid, sub_id) {
+      const key = subscriptionsKey(uid, sub_id);
```

Comment:
Some of the errors possible in the real implementation are not possible in the stub, will that be a hindrance to testing?

Reply:
Possibly, I can give this another pass to try to make it closer to my guesses at subhub

Reply:
On further thought, I think this might be worth expanding on as part of #719 when we're more certain about the actual error that will come out of subhub. Going to add a note to that issue.

Reply:
https://github.com/mozilla/fxa/issues/719#issuecomment-484695091